sourcerer-dbstore-jdbc/src/main/java/org/elder/sourcerer2/dbstore/jdbc/JdbcEventStore.kt

```diff
@@ -0,0 +1,395 @@
+    ): List<DbstoreEventRow> {
+        val readStatement = createReadStreamStatement(
+                streamId,
+                category,
+                effectiveMaxEvents,
+                fromVersion)
+
+        return readStatement.executeQuery().use {
+            it.readEventRowResults()
+        }
+    }
+
+    private fun Connection.createReadStreamStatement(
+            streamId: StreamId,
+            categoryName: String,
+            effectiveMaxEvents: Int,
+            fromVersion: DbstoreStreamVersion?
+    ): PreparedStatement {
+        var i = 1
+        val query = makeReadStreamEventsQuery(fromVersion != null)
+        val statement = prepareStatement(query)
+        statement.setString(i++, streamId.identifier)
+        statement.setString(i++, categoryName)
+
+        fromVersion?.let {
+            statement.setTimestamp(i++, Timestamp.from(fromVersion.timestamp))
+            statement.setTimestamp(i++, Timestamp.from(fromVersion.timestamp))
+            statement.setInt(i++, fromVersion.transactionSequenceNr)
+        }
+
+        statement.setInt(i, effectiveMaxEvents)
+        return statement
+    }
+
+    private fun <T> withConnection(readOnly: Boolean, dbOperation: (Connection) -> T): T {
+        return dataSource.connection.use {
+            it.autoCommit = false
+            var committed = false
+            if (readOnly) {
+                it.isReadOnly = true
+                it.transactionIsolation = Connection.TRANSACTION_READ_COMMITTED
+            } else {
+                it.transactionIsolation = Connection.TRANSACTION_SERIALIZABLE
+            }
+            try {
+                val result = dbOperation(it)
+                it.commit()
+                committed = true
+                result
+            } finally {
+                if (!readOnly && !committed) {
+                    it.rollback()
+                }
+            }
+        }
+    }
+
+    private fun getMaxReadBatchSize(requestedBatchSize: Int): Int {
+        return minOf(requestedBatchSize, maxReadBatchSize)
+    }
+
+    private fun ResultSet.readEventRowResults(): List<DbstoreEventRow> {
+        return this.readRowsWith(::readEventRow)
+    }
+
+    private fun readEventRow(row: ResultSet): DbstoreEventRow {
+        val streamId = StreamId.ofString(row.getString(1))
+        val category = row.getString(2)
+        val shard = row.getInt(3)
+        val timestamp = row.getTimestamp(4).toInstant()
+
+        return when (timestamp) {
+            SENTINEL_TIMESTAMP ->
+                // This is the end of stream marker, only extract the bits relevant to it
+                DbstoreEventRow.EndOfStream(
+                        streamId = streamId,
+                        shard = shard,
+                        category = category
+                )
+            else -> {
+                val transactionSeqNr = row.getInt(5)
+                val eventId = EventId.ofUuid(UUID.fromString(row.getString(6)))
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.elder.org/coding_rules#rule_key=detekt-kotlin%3AMagicNumber)